File: clouditor-engine-k8s/src/main/java/io/clouditor/credentials/KubernetesAccount.java
package io.clouditor.credentials;

import com.auth0.jwt.JWT;
import com.fasterxml.jackson.annotation.JsonTypeName;
import io.kubernetes.client.ApiException;
import io.kubernetes.client.apis.CoreV1Api;
import io.kubernetes.client.util.ClientBuilder;
import io.kubernetes.client.util.Config;
import io.kubernetes.client.util.credentials.AccessTokenAuthentication;
import io.kubernetes.client.util.credentials.KubeconfigAuthentication;
import java.io.IOException;

@JsonTypeName(value = "Kubernetes")
public class KubernetesAccount extends CloudAccount<ClientBuilder> {

  private String url;
  private String token;
  private String caCertificate;

  @Override
  public void validate() throws IOException {
    var builder = resolveCredentials();

    var api = new CoreV1Api(builder.build());

    try {
      var list = api.listNamespace(null, null, null, null, null, null, null, null);
      list.getItems();

      if (builder.getAuthentication() instanceof AccessTokenAuthentication) {
        // seems to work, lets decode the JWT to set username
        var decoded = JWT.decode(token);
        var claim = decoded.getClaim("kubernetes.io/serviceaccount/service-account.name");

        if (!claim.isNull()) {
          this.setUser(claim.asString());
        }
      } else if (builder.getAuthentication() instanceof KubeconfigAuthentication) {
        var kubeConfig = builder.getAuthentication();

        // TODO: maybe extract user from client cert, but it is hidden in a private field
      }

      this.setAccountId(api.getApiClient().getBasePath());

      LOGGER.info("Account {} validated with user {}.", this.accountId, this.user);
    } catch (ApiException e) {
      throw new IOException(e);
    }
  }

  /**
   * Discovers an Kubernetes account.
   *
   * @return null, if no account was discovered. Otherwise the discovered {@link KubernetesAccount}.
   */
  public static KubernetesAccount discover() {
    try {
      var account = new KubernetesAccount();

      // use the default client config
      var client = Config.defaultClient();
      var api = new CoreV1Api(client);

      var list = api.listNamespace(null, null, null, null, null, null, null, null);
      list.getItems();

      account.setAutoDiscovered(true);
      account.setAccountId(api.getApiClient().getBasePath());
      /*account.setUser(identity.arn());*/

      return account;
    } catch (IOException | ApiException ex) {
      // TODO: log error, etc.
      return null;
    }
  }

  @Override
  public ClientBuilder resolveCredentials() throws IOException {
    if (this.isAutoDiscovered()) {
      return ClientBuilder.standard();
    }

    return new ClientBuilder()
        .setBasePath(url)
        .setAuthentication(new AccessTokenAuthentication(token))
        .setVerifyingSsl(true)
        .setCertificateAuthority(caCertificate.getBytes());
  }
}

File: clouditor-engine-k8s/src/main/java/io/clouditor/credentials/KubernetesAccount.kt
package io.clouditor.credentials

import com.auth0.jwt.JWT
import io.kubernetes.client.apis.CoreV1Api
import kotlin.Throws
import java.io.IOException
import io.kubernetes.client.models.V1Probe
import io.clouditor.discovery.MixInIgnore
import io.clouditor.credentials.KubernetesAccount
import io.clouditor.data_access_layer.HibernatePersistence
import io.clouditor.discovery.ScannerInfo
import io.clouditor.discovery.k8s.KubernetesScanner
import io.kubernetes.client.models.V1Pod
import io.clouditor.discovery.ScanException
import io.kubernetes.client.models.V1PodList
import io.kubernetes.client.ApiException
import io.kubernetes.client.models.NetworkingV1beta1Ingress
import io.kubernetes.client.apis.NetworkingV1beta1Api
import com.fasterxml.jackson.annotation.JsonTypeName
import io.clouditor.credentials.CloudAccount
import io.kubernetes.client.models.V1NamespaceList
import io.kubernetes.client.util.credentials.AccessTokenAuthentication
import com.auth0.jwt.interfaces.DecodedJWT
import com.auth0.jwt.interfaces.Claim
import io.kubernetes.client.util.credentials.KubeconfigAuthentication
import io.kubernetes.client.ApiClient
import io.kubernetes.client.util.ClientBuilder
import io.kubernetes.client.util.Config

@JsonTypeName(value = "Kubernetes")
class KubernetesAccount : CloudAccount<ClientBuilder?>() {
    private val url: String? = null
    private val token: String? = null
    private val caCertificate: String? = null
    @Throws(IOException::class)
    override fun validate() {
        val builder = resolveCredentials()
        val api = CoreV1Api(builder.build())
        try {
            val list = api.listNamespace(null, null, null, null, null, null, null, null)
            list.items
            if (builder.authentication is AccessTokenAuthentication) {
                // seems to work, lets decode the JWT to set username
                val decoded = JWT.decode(token)
                val claim = decoded.getClaim("kubernetes.io/serviceaccount/service-account.name")
                if (!claim.isNull) {
                    setUser(claim.asString())
                }
            } else if (builder.authentication is KubeconfigAuthentication) {
                val kubeConfig = builder.authentication

                // TODO: maybe extract user from client cert, but it is hidden in a private field
            }
            setAccountId(api.apiClient.basePath)
            LOGGER.info("Account {} validated with user {}.", accountId, user)
        } catch (e: ApiException) {
            throw IOException(e)
        }
    }

    @Throws(IOException::class)
    override fun resolveCredentials(): ClientBuilder {
        return if (this.isAutoDiscovered) {
            ClientBuilder.standard()
        } else ClientBuilder()
            .setBasePath(url)
            .setAuthentication(AccessTokenAuthentication(token))
            .setVerifyingSsl(true)
            .setCertificateAuthority(caCertificate!!.toByteArray())
    }

    companion object {
        /**
         * Discovers an Kubernetes account.
         *
         * @return null, if no account was discovered. Otherwise the discovered [KubernetesAccount].
         */
        fun discover(): KubernetesAccount? {
            return try {
                val account = KubernetesAccount()

                // use the default client config
                val client = Config.defaultClient()
                val api = CoreV1Api(client)
                val list = api.listNamespace(null, null, null, null, null, null, null, null)
                list.items
                account.isAutoDiscovered = true
                account.setAccountId(api.apiClient.basePath)
                /*account.setUser(identity.arn());*/account
            } catch (ex: IOException) {
                // TODO: log error, etc.
                null
            } catch (ex: ApiException) {
                null
            }
        }
    }
}